File: Data Structure I/5. Tree/P700.cpp
// Date: Sun, Nov 21, 2021
// 700. Search in a Binary Search Tree

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	// depth-first search approach
	// time complexity: O(n)
	// space complexity: O(1)
    TreeNode* searchBST(TreeNode* root, int val) {
    	if (!root)
    		return nullptr;
        if (root->val == val)
        	return root;
        if (root->val < val)
        	return searchBST(root->right, val);
        return searchBST(root->left, val);
    }
};

File: LeetCoding-Challenge/April-2022/P700.java
package leetcodingchallenge.april2022;

/**
 * Date: April 14, 2022
 * 700. Search in a Binary Search Tree
 * https://leetcode.com/problems/search-in-a-binary-search-tree/
 */

public class P700 {
    // time: O(log(n))
    // space: O(1)
    public TreeNode searchBST(TreeNode root, int val) {
        TreeNode current = root;
        while (current != null) {
            if (current.val == val) return current;
            current = current.val > val ? current.left : current.right;
        }
        return null;
    }

    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}
